Merge of Modules branch into main

luketpeterson opened the pull request:
There is a lot to say here.  Probably best to read the `modules` & `pkg_mgmt` documentation here:

https://github.com/luketpeterson/hyperon-experimental/blob/modules/docs/modules_dev.md
and
https://github.com/luketpeterson/hyperon-experimental/blob/modules/docs/modules_internal_discussion.md

Necr0x0Der commented:
1) While `f1_imports.metta` seems to work in the test environment, imports of local metta scripts don't seem to work when I'm trying to run `f1_imports.metta` as a separate script either by REPL or by metta.py . It also doesn't seem to work in settings in which `extend-py!` works for other projects. This is inconvenient. Can it be fixed?
2) Some projects use `load_py_module` directly. Should it be replaced with `import_dependency`? It's difficult to check now because of (1)

luketpeterson commented:
> 1. While f1_imports.metta seems to work in the test environment, imports of local metta scripts don't seem to work when I'm trying to run f1_imports.metta as a separate script either by REPL or by metta.py . It also doesn't seem to work in settings in which extend-py! works for other projects. This is inconvenient. Can it be fixed?

Regarding the `metta.py` script, I think the issue was that the runner didn’t know to search in the directory next to the loaded script.  I fixed that in e5d8ed4

regarding the REPL, can you be more specific about what you tried and what you expect.  It works for me when I run:
```
./hyperon-experimental/target/debug/metta ./hyperon-experimental/python/tests/scripts/f1_imports.metta
```

NOTE: the `hyperon-experimental/repl/target/` directory is deprecated as of the build system changes from a few weeks ago, so it should be removed.  All targets are now built into `hyperon-experimental/target`, regardless of the sub-project.

> 2. Some projects use load_py_module directly. Should it be replaced with import_dependency? It's difficult to check now because of (1)

Loading and importing are separate operations.  The MeTTa `import!` op will use the catalogs to load a module if one isn't already loaded, but you can also load modules manually.

If the goal is to load a specific module from a known location in the file system, `MeTTa.load_module_at_path` is a direct way to do that.  It will load a python module or a `.metta` module or any other valid format, depending on what the path points to.

If people want to import python modules from arbitrary locations in the file system, first load them with `MeTTa.load_module_at_path`, and then import them from MeTTa using `import!`.

`import_dependency` is used to import a loaded module into another running module, but it's a method of RunContext, so it's only available within a runner thread.  It's the same code path called by the `import!` operation after the name has been resolved.  It was mainly designed for use in the implementation of loader functions and grounded execution.

Necr0x0Der commented:
> Regarding the metta.py script, I think the issue was that the runner didn’t know to search in the directory next to the loaded script. I fixed that

Thanks.

> NOTE: the hyperon-experimental/repl/target/ directory is deprecated 

Thanks for this note. It resolves my issue.

On additional notes / issues:

1) I see a number of places in this repo, where `extend-py!` is still used. Some of them are in sandbox examples and in integration tests. While it is not that important, but it would be convenient to fix them in this PR (maybe, except sandbox/resolve, which I need to review whether it becomes deprecate or not). Others are in metta_shims.rs, which may need some fixes as well?

2) We have some issues with metta-motto (besides just renamed interface functions), which would be very nice to fix before merging, because metta-motto is now used by four different dev groups. While importing `motto` module itself works now (after fixing the runner environment), metta-motto uses `import_file` internally, and replacing it with `load_module_at_path` (which seems to work instead of `import_file` in unit tests now) results in panic `Fatal Error loading basic_agent.msa. Module without a working directory cannot load dependencies by relative path` (catalog.rs:255:39). JIC, replacing `.msa` with `.metta` doesn't solve the issue. My guess is that the fix for the runner you introduced in `metta.py` would be nice to introduce to `MeTTa()`. metta.py is not fundamental, while `MeTTa` class is.

luketpeterson commented:
> 1. I see a number of places in this repo, where extend-py! is still used.

I replaced all of them in the https://github.com/trueagi-io/hyperon-experimental/pull/580/commits/8aa70c9ae7a4bd2a96fb92c7bc5e9d61ef40f734 commit, and also added some improvements to make it easier to debug what happened when a module fails to load.

One tiny thing concerns me: the `sandbox/resolve` project contains a file called `r.metta` and `r.py`.  In this case it's fine because the Python format loader checks first, so there is no circular import when `r.metta` imports the `r` module, and it resolves to `r.py`.  But since people are likely to use the sandbox and examples as sample code, I think we should illustrate the best practice of keeping the module names distinct.

> 2. We have some issues with metta-motto (besides just renamed interface functions)

How can I test this myself?  I added a fallback to use the os cwd for explicit file loads (but not as part of the automatic resolution) which might fix this case.  But without seeing the problem myself I don't know.

> My guess is that the fix for the runner you introduced in metta.py would be nice to introduce to MeTTa()

In `metta.py` I set the runner's working_dir based on the location of the metta file being run.  But when a MeTTa object is being initialized we don't know where the file being run is (or if the code comes from a file at all).

One possible idea is to change the default runner init code to use the current cwd if no working_dir is explicitly provided, (or explicitly disabled, which is also an option)

Necr0x0Der commented:
> I replaced all of them

Thanks

> One tiny thing concerns me: the sandbox/resolve project contains a file called r.metta and r.py

As I said, resolve example should be revisited. However, I'd like to make two comments:
1) resolve example includes calling methods in the imported module. Does the new module system already supports this? It's ok if not especially since Minimal MeTTa has a space as an explicit argument for interpreting expressions
2) In metta-motto, we have changed extensions for metta-scripts with specific purposes. I believe providing file names with extensions can be useful, and it can help disambiguating `r.metta` and `r.py` as well. There are different ways to support extensions, though. Apparently, the story can be different for metta and python files (although we did have the support for providing paths to python files/modules in the previous version; idk how it is handled here). Also, the strategy can be different for `import!` in metta-files and for the interface functions like `load_module_at_path` (which, for example, can have an optional flag explicitly indicating that the file should be loaded as metta).

> In metta.py I set the runner's working_dir based on the location of the metta file being run. But when a MeTTa object is being initialized we don't know where the file being run is (or if the code comes from a file at all).

Yes, in this case it should be the folder of the python script. It's indeed different from `metta.py` case.

> One possible idea is to change the default runner init code to use the current cwd if no working_dir is explicitly provided, (or explicitly disabled, which is also an option)

It would be nice to set it to the main Python script folder, but maybe it's not the job for `MeTTa()`. Using cwd when `MeTTa()` is called without explicit `working_dir` sounds good. It's the caller issue to call scripts with the right cwd if they rely on it, or to set up the environment with the necessary include paths. So, let's go for cwd.

> How can I test this myself?

I'm refactoring metta-motto in a branch, and will push this branch to github today.

Necr0x0Der commented:
As I understand, cwd was already added. I'm still getting the error
`[2024-02-13T08:21:15Z ERROR hyperonc::util] Python error loading MeTTa module '02_metta_agent'. AttributeError("'NoneType' object has no attribute 'loader'")`
even with the right cwd. So I wonder what can be the reason for this error.

Necr0x0Der commented:
One possible issue is that `import_file` returned the result of evaluation of the script. Now, it returns `ModuleId`. How to get the result of evaluation of the module/script?

luketpeterson commented:
> ... How to get the result of evaluation of the module/script?

My thinking here was that a module is different from just inline code execution.  So a module loader either returns success or failure.  If you want to move additional information out of the module, the module can define some atoms to carry that information and make it available to the client.

But if it's just inline code execution you want, then `MeTTa.run` is a better fit all around.

> but I still wonder if there is a solution from the module system side. [The reason why I wouldn't like to load the file is to avoid solving the path search problem if it is handled by MeTTa module system already]

Is the idea that the files you want to load are in the module's directory, so you know where the files are relative to that directory, but not the path to the directory itself?  That is the idea of a module resources directory (called the module working directory in some places, but I think it should be renamed to be consistent and avoid confusion with the runner's working dir)

What do you think of a `run-file!` MeTTa operation that treats a symbol argument as a filename, and loads it from the resource directory and runs it?  Does that give you want you are looking for?

> ...`basic_agent.msa`...

This has given me some pause because `basic_agent.msa` would be parsed as a two-level module name because the '.' character is a separator.  But it seems that this is one module name in your use.

I think I will change the separator to something else because others will certainly hit the same issues.  What do you think of ':' or "::"?  I think I'm slightly leaning towards single ':' because it's less likely to conflict with other uses.  For example, in the resolve example, there is a token regex that would conflict with "::", but people are less likely to redefine a single character.

Necr0x0Der commented:
> My thinking here was that a module is different from just inline code execution. So a module loader either returns success or failure. If you want to move additional information out of the module, the module can define some atoms to carry that information and make it available to the client.

This makes sense in general, but this doesn't solve the issue that the functionality which was useful is now lacking. I can agree that loading a module is different from running a script, but since the module can contain executable stuff, the result of this execution can appear to be useful in some cases. Just discarding it is wasteful. Maybe, we can have another API function like `load_module_at_path_with_result` or something else...

> Is the idea that the files you want to load are in the module's directory, so you know where the files are relative to that directory, but not the path to the directory itself?

It's also a possibility, although I'm just using `import_file`, because it worked exactly as I needed. But I can load scripts manually and use `MeTTa.run` for now and see if I encounter any problems with path resolution

> What do you think of a run-file! MeTTa operation that treats a symbol argument as a filename

I needed a Python API function, not a function in MeTTa. It's temporarily not an issue.

However, while metta-motto agent scripts are now loaded and executed explicitly, I got a more problematic issue. I have prompt templates, which are also metta scripts, and they are imported in metta. It causes the error `Illegal module name: basic_script.mps`. But if I rename `.mps` to `.metta`, there is still a problem that stdlib (and corelib) is now loaded into the module space, and the loaded module is not aware of tokens of the parent module. Both are ok, but the question is how to achieve the former functionality

Necr0x0Der commented:
@luketpeterson , what is the alternative to `load_py_module` now? I thought it should be `load_module_at_path`, but it doesn't work for Python modules (formerly, extensions): when `!(import! motto &self)` works, `load_module_at_path('motto')` doesn't.

Necr0x0Der commented:
I have similar issues not only with Python modules, but also with importing metta scripts. I'd really like to have an equivalent in Python API, which would work similar to `!(import my-module &self)`, but with the possibility to specify a relative path or custom file name.
I added some custom script loading in metta-motto to work around the changes in the module system, and it is starting to work again ( https://github.com/zarqa-ai/metta-motto/pull/37 ), but still the reduced Python API is not convenient, and some stuff is still not working. 
Two particular things:
1) It would be convenient to use `load_module_at_path` or another function in tests/test_scripts.py . It can possibly be done by the environment builder, but why couldn't it work like it's now there? I also replaced `load_py_module("motto")` with `m.run('!(import! motto &self)')` in tests/test_python_direct.py , but once again, it would make sense for a Python function to work here instead of executing metta code.
2) In tests/test_sparql_select.py , we'd like to import a submodule. It works neither with `load_module_at_path` nor with `import!`. I'm still getting `Illegal module name`.

luketpeterson commented:
> It causes the error `Illegal module name: basic_script.mps`

I have made sure '.' is now a legal character in a module name, and relaxed the loader logic so that it does't require modules to be named '.metta'.  So the first layer of problems with different file extensions should be better with https://github.com/trueagi-io/hyperon-experimental/pull/580/commits/0f4a0f4f9d44a19e267b6875f43b9f5bd5d17fb3

> the loaded module is not aware of tokens of the parent module

This is by design.  Dependencies only flow in one direction.  If the parent (I call it client in the docs) tokens can affect the sub-module then in effect the sub-module isn’t modular; it’s more like a continuation of the parent module that can’t exist by itself.  If you want some tokens to be available to sub-modules, they need to be defined in yet another module that is imported by everyone who needs them.

However, It sounds to me like these “non-modules” are an important part of your use case, so i want to find an API will give you (and others with a similar use case) the functionality you need in the most elegant way.

It sounds like you are looking for something more along the lines of “#include” (with C semantics) than `import` / `use` (with Rust or Python semantics (I know Python has both)), but it sounds like you still want the package-management features. (search, versioning, remote catalogs, etc.)

After thinking about this for the past few days, one idea is to add a `include!` operation that is different from `import!`.  To make this work I would need to generalize the catalog system to be able to search for any resource, not just MeTTa modules.  Then `import!` could check the catalog for a text blob, and just parse the text in the context of the running module and inline the results at the call site.

It will take me a day or so to prototype this because the catalog was designed with modules in mind.  But I think generalizing the mechanism to other resource types might be a good idea in general.  For example, we could host NN model files or whatever, through the same mechanism, without needing to wrap the resource in a MeTTa module that does nothing.

> `!(import! motto &self)` works, `load_module_at_path('motto’)` doesn't

`import!` takes a module name (or module namespace path) as its argument.  It goes through the 3 stages of resolve->load->import, whereas `load_module_at_path()` only performs the `load` part.  I have created a diagram to describe the behavior of `import!` here: https://github.com/luketpeterson/hyperon-experimental/blob/modules/docs/assets/importop_behavior.svg which I will add to the documentation soon.

`load_module_at_path()` takes a file-system path as an argument.  It’s either an absolute path or a relative path, relative to the runner’s working dir.  It bypasses all search logic because it is the API used when you want to load a very specific file / module, as opposed to loading the best available module given the available catalogs.  (include directories are catalogs, but so are servers, etc.)

> what is the alternative to `load_py_module` now?

The semantics of `load_py_module` don’t have an exact equivalent because those semantics were based on a runner design that fundamentally didn’t have modularity, and didn’t distinguish modules from the files they were loaded from.

Regarding loading modules from Python, it depends on

1. how you want to identify the module you are looking for.

If you want to resolve a module name, use `RunContext.load_module()`  This would be done for shared modules.  Think “numpy” or “PyTorch”.  Or, if you have an exact file system path then use `load_module_at_path()`  This would be done for an internal sub-module in your module’s resource dir, or for a quick-and-dirty dependency like a test or a component of a larger app.

Because the name resolution behavior depends on the running module (see https://github.com/luketpeterson/hyperon-experimental/blob/modules/docs/assets/module_name_resolution.svg ), `load_module()` is a `RunContext` method.  ie. it’s designed to be called from inside a runner thread, not from the outside of the runner like a `MeTTa` object method.

2. where you want to import the module into.

The `RunContext.import_dependency()` method performs the importing part of the `import!` op (^1), but it’s also a `RunContext` method.  This is because it would otherwise be ambiguous as to the meaning of `&self`.

There is a Rust method to run an arbitrary function within a specified context, but I didn’t bridge it into the C and Python APIs yet.  I’ll do that today and reply with the Python code to mimmic the behavior of `import!` without any MeTTa code, called from outside a runner.

^1: Assuming you want `import *` aka (import! mod_name &self) behavior.

Necr0x0Der commented:
@luketpeterson , the necessity to have functions to deal with modules and pure spaces separately was discussed long time ago. But it is not urgent, since I already tweaked the way scripts are loaded in metta-motto. No need to spend time on this right now. As for `load_module_at_path`, thanks for explanations why it doesn't work in the same way as `import!`, but what I asked is a Python API function which would work as `import!`. But I can just load a file and run it. It's also not a blocker.
What is the urgent blockers is that I'm still unable to `!(import! motto:sparql_gate &self)`

Necr0x0Der reviewed (approved):
I just switched to the flat module system for metta-motto. So, I'm ok merging it now. While there are many things to improve, the earlier we merge the faster we'll start getting complains from other metta-users.

luketpeterson commented:
> but what I asked is a Python API function which would work as `import!`

I'll add this shortly.  The API exists in Rust but I didn't add it to C and Python yet.

> What is the urgent blockers is that I'm still unable to !(import! motto:sparql_gate &self)

I made the change to migrate from a flat module namespace to a hierarchical namespace over the weekend, but I'm still rounding out the behaviors.  Specifically I need to make sure recursive loading works, and currently it seems that it doesn't work in all cases.

patham9 commented:
What is the planned timeline of this PR? Are these changes planned before the alpha release? I guess after as it looks like a major breaking change which affects imports and could hence induce some delays?

vsbogd commented:
> What is the planned timeline of this PR? Are these changes planned before the alpha release? I guess after as it looks like a major breaking change which affects imports and could hence induce some delays?

@patham9 we are going to include this changes into alpha release thus merge it before it. We will have a separate release before merging this. For example I have made https://github.com/trueagi-io/hyperon-experimental/releases/tag/v0.1.6 recently.

I am not sure which delays do you mean? We need to list the list of breakages this PR introduces to make it better understand what is a migration path for the users.

patham9 commented:
@vsbogd: Thank you, this makes perfect sense!
Regarding delays: If one can get away with replacing extend-py! with the new import statement, without needing to change the Python files which one wants to import themselves, then it should be quite quick to adapt. However if some extensive modifications are required to the import files themselves, the people who are working to complete their components for the alpha release will have some unexpected extra workload which will likely cause delays. For instance all MeTTa code I wrote so far works with the Rust MeTTa interpreter of August 2023 as well as the one of February 2024, this PR however seems like it could be a breaking change which hopefully will only require minor work to adapt to.

Necr0x0Der commented:
My intention was to merge it ASAP with some follow-up improvements in separate PRs because it's blocking further work on modules, but maybe @vsbogd and @luketpeterson would prefer another strategy

Necr0x0Der commented on lib/src/metta/runner/stdlib.rs:
```diff
@@ -61,101 +47,79 @@ impl Display for ImportOp {
 
 impl Grounded for ImportOp {
     fn type_(&self) -> Atom {
+        //QUESTION: How do we express that we can accept a variable number of arguments?
+        // Ideally the "import as" / "import into" part would be optional
```
Yes, I'm simulating variable number of arguments wrapping them into `Expression`. Not sure if non-deterministic types is a good idea here. They produce additional results sometimes

luketpeterson replied:
> they will produce additional results if arguments can be casted to more than one type.

If we take any single Atom as the arg, and then do type checking inside the grounded atom's execute function that seems workable.

Are you saying the grounded atom's execute function might be invoked multiple times?  Because if that were true that would also be true for any signature.

@vsbogd @Necr0x0Der Do you think we should change the `import!` signature now?  Since changing it later will have downstream effects on all users?

Or should we wait until after alpha / next-breaking-revision to see where we stand and decide on how we want to support more import options?

vsbogd replied:
> Are you saying the grounded atom's execute function might be invoked multiple times? Because if that were true that would also be true for any signature.

It is true for any function grounded or not. It works like I wrote it is called many times if arguments can be casted to more than one type.

vsbogd replied:
> @vsbogd @Necr0x0Der Do you think we should change the import! signature now? Since changing it later will have downstream effects on all users?

Am I right this PR already changes a signature (args of import! are swapped)? Do we want to keep it or swap them back before merging? If we change signature what kind of imports will be supported automatically?

Necr0x0Der replied:
Yes, args of `import!` are swapped, so the signature is changed. If we are thinking about extended API for `import!`, it would possibly be better to make the signature suitable for further changes or to undo this change for backward compatibility (if we are thinking about changing it again, we should avoid two changes, and both of these approaches can satisfy this).
One question I have is about evaluating functions in the context of the imported space as well as accessing tokens of the imported module. Depending on this, we either need a way to assign names to imported modules or not.
I don't like too much how it was done originally. It was convenient for a narrow use, but it raises some questions. Currently, `import!` adds a new token if the argument is not `Space`. It's convenient for some cases, but I'd prefer `import` to just return a space in others (although it wouldn't work well for modules). it's also a little bit weird for namespacing, especially if we want to use tokens defined in another module. If a module has both tokens and non-empty space content, they if we import it to another space, we lose access to tokens. If we import it to `&self`, we mix up its content with `&self`. That's why I proposed to separate imports of spaces from imports of modules. Module is a space with a tokenizer, or, more generally, is a runner. We can have, say, rholang as a whole runner - not just space. And we already have something like this for metta-morph, although in a custom ad hoc way. We need something like Runner API besides Space API...
As for,
```
!(import! (<name> into <space>)) ; import specific name into space
!(import! (<name> as <alias> into <space>)) ; to import into space as alias
!(import! <name>) ; to import into self
!(import! (<name> as <alias>)) ; to import into self as alias
```
I don't like `into` too much as it might be excessive. I don't like `!(import! <name>)` too much, because importing a space to `&self` shouldn't be a good practice in general. And I'm not sure how `alias` should be used. And we go back to the question about use cases. We need extensions typically for additional tokens, which we'd like to access easily. Sometimes, with spaces. We can have a special mechanism for importing tokens, e.g. `(from das_gate import new-das)` or `(from das_gate import *)` . But it will be confusing that we cannot import normal functions from modules this way. We can have `import!` and `extend!`, which behave differently (`extend!` will import all tokens to the caller module even if the module is imported as a separate space). Extensions being imported to `&self` can actually create a space with their additional code in any case and provide access to this code space via a special token they create by themselves. But it would be better handled by the module system in a uniform way, I think. 

Thus, there is a lot to discuss. I wouldn't want these considerations and issues to be a blocker for merging this PR. Maybe, we need to just swap arguments of `import!` back, and do the next major revision of imports before alpha release.

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+//!
+
+//LP-TODO-NEXT make a test to make sure circular imports are caught and don't lead to infinite recursion
+//QUESTION: Should circular imports between modules be allowed?  The current implementation (and possibly
+// the MeTTa language itself) disallow circular imports because there is no concept of forward declaration.
+// It *may* be possible to pre-parse the language in order to make recursive imports possible, but I have
+// not yet thought in detail about this.
```
The current state of the code isn't great because loading a module on-demand can still cause problems if a sub-module is referenced by a module loader, before the module has finished initialization.

I think I need to spend a few hours on this, to make sure I have tested all the edge cases, and write unit tests.  But it's not a block-merge.

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+// not yet thought in detail about this.
+//
+
+//QUESTION on shared base dependencies & sat-set solving:
+//The currently implemented design resolves each module's dependencies in a straightforward depth-first
+//  order.  This is possible because the module system allows multiple instances of the same module to
+//  be loaded simultaneously.  So each module can pick its best dependencies based on its pkg-info and
+//  the available catalogs.
+//However, consider the following situation:
+//  ModA depends on ModI for some interface types (for example a special String type)
+//  ModB depends on ModI for the same interface types, but ModA and ModB don't know about each other
+//  ModTop depends on both ModA and ModB, and uses functionality in ModA to create some objects that
+//   it expects ModB to be able to use.  Therefore the system must guarantee the same version of ModI
+//   gets imported by both ModA and ModB.
+//This is precisely the opposite behavior from the ability of a module to carry around "private"
+//  dependencies and know that those dependencies will always be loaded, and they won't be substituted
+//  for another version.
+//
+//I see several possible solutions:
+// 1.) We could disallow private dependencies altogether.  This is the approach taken by Cargo.
+//  However this contravenes some of the desiderata outlined in this issue:
+//  https://github.com/trueagi-io/hyperon-experimental/issues/470
+// 2.) We could require explicit re-exporting of a dependency module used in a module's interface, which
+//   would give the implementation an opportunity to find dependency module versions that work for
+//   all other modules that use them in common.  ie. solve for the sat set.  Optionally, with this approach,
+//   the module could also opt to re-export a private dependency as part of itself, making the interface
+//   between ModA and ModB in the example deliberately incompatible. 
+// 3.) We could require private dependencies to be explicitly specified as private in the pkg-info.  With
+//  the default behavior being a single module for each module name.  This might be a reasonable compromise
+//  between 1 & 2, however we would likely need some form of linting, so that a user doesn't shoot
+//  themselves in the foot by exporting an interface that includes items from a private dependency
+//
+// I think my personal preference is for #2.
```
Agreed RE dependency versioning.  The parent (client) that uses both sub-modules is in a position to find the transitive sub-module versions that are mutually compatible.  So we can kick the sat-solver into the future.

When we have a larger ecosystem this might be demanded, but it's probably years in the future

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+/// `ModuleCatalog` types are closely connected with [ModuleLoader] types because the `ModuleCatalog` must
+/// recognize the module in whatever media it exists, and supply the `ModuleLoader` to load that module
+//
+//QUESTION: Should it be legal for a catalog to alias a module?  In other words, can a catalog's `lookup`
+// method return a ModuleDescriptor where the name field of the ModuleDescriptor doesn't match the query
+// name passed to `lookup()`?
+//
+// * Arguments *Against* it being legal:
+//  If it is allowed, we have destroyed the direct mapping between loaded modules and the names they
+//  have in the import! operation.  In other words, if we start with a ModuleDescriptor, and import!
+//  its name field, there is no guarantee we will get the same module at all.
+//
+//UPDATE: I started by writing a balanced view of the pros and cons, but ultimately I found the arguments
+// in favor to be weak, and potentially anti-features.  So I think it should be illegal to return a
+// ModuleDescriptor where the name field doesn't match the `lookup` name.
+//
+//UPDATE2: Aliases from multiple names in the runner's namespace to a single module are now allowed, but
+// but a module's name from its ModuleDescriptor will always point back to that module.
+//UPDATE3: I think this is ultimately resolved by removing the name method from a module loader.  Then:
+// 1.) The name a module gets in the namespace depends on the name it was loaded with
+// 2.) The module loader can no longer conflict with that, and the module descriptor is irrelevant once
+//   the module is loaded
+// Therefore, it's up to the catalog to make sure relevant modules are provided, but if the catalog
+// has a good reason to alias a module, it will now be able to do it.  I will implement this shortly,
+// which will involve deleting this comment altogether.  So this update is part of the "historical record"
```
Agreed.  I will implement this very shortly.

luketpeterson replied:
Implemented in https://github.com/trueagi-io/hyperon-experimental/commit/8a0798e028e9ac074b6f4cc67223cf884aac5038

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+    /// Indicates that the dependency module should be loaded from a specific FS path
+    ///
+    /// If the fs_path is specified, the other pkg_info attributes will be ignored.
+    //QUESTION: We need a MeTTa "style guide" for these field names, since they are effective going
+    // to be part of the API, because a PkgInfo will be deserialized from atoms
+    pub fs_path: Option<PathBuf>
```
This is related to the `atom-serde`; and before I implement `atom-serde` I'll probably use Hjson or TOML (which do you prefer?) to encode a pkg-info file for a directory-based module format.

But basically it's a style question.  What capitalization / style scheme should we use for the keys that are user-facing?  ie. `fsPath` vs `fs_path` vs. ???

And more generally, I think we are getting to the point we should be thinking about a MeTTa style guide so people know how to quickly interpret what kind of an object they have based on the capitalization / style scheme.

vsbogd replied:
>  I'll probably use Hjson or TOML (which do you prefer?) 

I didn't hear about Hjson yet and I used TOML only recently. Both formats are not as wide as JSON but both are human readable/writable and supported. I don't have a strong opinion here.

> What capitalization / style scheme should we use for the keys that are user-facing? ie. fsPath vs fs_path vs. ???

In MeTTa code I trying stick to Scheme code style (https://www.r6rs.org/final/html/r6rs/r6rs-Z-H-2.html#node_toc_node_sec_6.7) but I don't think it is the only option.

luketpeterson replied:
Hjson is nice because most people think it's just JSON.  Basically it is the JSON that the javascript interpreter will accept.  As opposed to the oppressively obnoxiously strict JSON of the JSON spec.

Personally I like Hjson quite a bit.  But TOML would also be fine.

> I trying stick to Scheme code style (https://www.r6rs.org/final/html/r6rs/r6rs-Z-H-2.html#node_toc_node_sec_6.7)

'-' as a separator is fine with me.

I'm a little confused about capitalization recommendations however.  I'm not sure I'm reading it correctly, but is Scheme case-insensitive?  The document said said "The parameter names are case-insensitive: Vector is the same as vector." But I may have over-generalized that sentence.  Does Scheme have capitalization recommendations?

vsbogd replied:
> Does Scheme have capitalization recommendations?

I am not sure actually. But our atoms are case-sensitive and we can make our own suggestions.

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+    pub fs_path: Option<PathBuf>
+
+    //TODO: field for fetching from a specific git repo
+    //TODO: field to indicate acceptable version range for dependency
```
Right.  PkgInfo will contain the version of the current package as well as version requirements for dependencies.  If there is no key then it's an un-versioned module.

When I write `atom-serde` any block of MeTTa code will be able to declare certain atoms that will embed the value.  Until then, the PkgInfo will need to come from somewhere else like a `_pkg-info.json` file in a directory-based module.

vsbogd replied:
> When I write atom-serde any block of MeTTa code will be able to declare certain atoms that will embed the value.

Not clean to me how is  `atom-serde` connected to this, but probably we need to discuss it somewhere else.

luketpeterson replied:
> Not clean to me how is atom-serde connected to this, but probably we need to discuss it somewhere else.

Sure.  It's definitely not a merge-blocker.  But the connection has two-levels of indirection.  1.) supplying a `pkg-info` is how certain module formats will express their version.  2.) The only way to do that in a module that's a single `.metta` file is with atoms in the MeTTa code, and those atoms will be interpreted as a `PkgInfo` struct using `atom-serde`.

luketpeterson commented on lib/src/metta/runner/modules/catalog.rs:
```diff
@@ -0,0 +1,568 @@
+impl ModuleCatalog for DirCatalog {
+    fn lookup(&self, name: &str) -> Vec<ModuleDescriptor> {
+
+        //QUESTION: How should we handle modules with an internal "package-name" that differs from their
+        // name in the file system?
+        //
+        //Cargo treats the module's internal "package-name" as authoritative, but that means it's impossible
+        // to "install" a module simply by dropping it into a directory in the search path, because there
+        // needs to be an index of all available modules in that directory.
+        //
+        //For us, I think we want a less formal approach akin to Python's, where we are allowed to drop a
+        // module into a directory, and import it with a naked `import` statement (i.e. no pkg-info entry)
+        // but for that to work, we need to stipulate that it's possible to infer a file name from a
+        // module name.
+        //
+        //NOTE: This is not a limitation across all catalogs, just the `DirCatalog`  If a catalog is able
+        // to maintain its own index of module names, it can store the modules any way it wants to.
+        //
```
This also partially becomes a moot point because of the idea that a `ModuleLoader` won't provide a `ModuleDescriptor` anymore.  So there won't be two places that provide names and thus no opportunity for them to conflict.

I'll address this at the same time I implement the fix for https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1491205886

luketpeterson commented on lib/src/metta/runner/modules/mod.rs:
```diff
@@ -0,0 +1,633 @@
+    ///     symbol that doesn't perfectly match `name`, a new [Tokenizer] entry will be created to
+    ///     access the atom in the &self module
+    ///
+    // QUESTION: This behavior of exactly matching a regex makes importing a Tokenizer pattern pretty
+    // unfriendly.  Does it make sense to require Tokenizers entries to be associated with atoms, for
+    // example "Type Atoms"?  For example, we could have an "Number" type that is tied to all the
+    // Tokenizer regex patters used to parse different types of numbers?  Then a user could
+    // "!(import! Number from Arithmetic)" or whatever, and get all the Tokenizer patters that parse
+    // numbers?
```
I should have linked https://github.com/trueagi-io/hyperon-experimental/issues/510 in the code, since that's basically a discussion on this point.

luketpeterson commented on lib/src/metta/runner/modules/mod.rs:
```diff
@@ -0,0 +1,633 @@
+    /// WARNING: Module import behavior is still WIP, specifically around "import *" behavior, and
+    /// especiallu around transitive imports
+    //
+    //QUESTION: What do we do about tokenizer entries?  Currently they end up glomming together but this
+    // is highly undesireable for several reasons.  At best it leads to tokenizer entries that are
+    // unreachable because they're superseded by other entries, and at worst this can introduce some
+    // difficult to diagnose behavior
```
They're related.  A more thorough treatment of these issues is in the `modules_internal_discussion.md` file, in the "## Question Topic 1: Importing & Linking :: Import All" section.  I have deleted this block of comments as the text in the .md file is more representative of the current state of understanding.

luketpeterson commented on lib/src/metta/runner/modules/mod.rs:
```diff
@@ -0,0 +1,633 @@
+    // unreachable because they're superseded by other entries, and at worst this can introduce some
+    // difficult to diagnose behavior
+    //
+    //QUESTION: How should we prevent duplication of transitive imports?  For example, consider: ModA
+    // imports ModOne and ModTwo.  ModOne imports ModB.  ModTwo also imports ModB.
+    // How do we make sure ModA doesn't end up with duplicate definitions for ModB?
+    //
+    //The solution implemented here is to elevate any transient dependencies up, and import them directly
+    // into the upper module, but this has a lot of drawbacks described in the [stripped_space] method
+    // comments, such as requiring a deep-copy of the dependency module's atoms.
+    //
+    //This is usually solved in other languages using some combination of strategies.
+    // 1.) Keep loaded modules contained within their own name-spaces. And allow aliasing between a parent's
+    //  space and a child module's space.  This is not really sufficient for us because we also want to
+    //  import Tokenizer entries and their associated functionality, such as the ability to parse integers
+    // 2.) Allow a conditional "import-unique" brute-force import which is tantamount to inlining the dependency
+    //  at the point of the import statement.  e.g. `#include` in C., but with the "ifdef" guards to make sure
+    //  the same header isn't inlined multiple times.  We may want to offer this functionality, but I assume
+    //  we'll want to offer this in addition to a more hygenic module system structure, given that most languages
+    //  (Python, JavaScript, even Perl!) that started off with brute-force imports have retrofitted more
+    //  sophisticated dependency management as they have matured
+    // 3.) Force a module to be explicit about what can be exported from the module.  e.g. the `pub` visibility
+    //  qualifiers in Rust, etc.
+    //
+    //Personally, I feel like 3. (being explicit about exported items) is the cleanest solution and provides
+    //  the most flexibility into the future (such as the ability to unload modules, etc.)
+    //
```
[this comment](https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1492011828) applies here too.

luketpeterson commented on lib/src/metta/runner/stdlib.rs:
```diff
@@ -27,20 +23,10 @@ fn unit_result() -> Result<Vec<Atom>, ExecError> {
 #[derive(Clone, Debug)]
 pub struct ImportOp {
-    metta: Metta,
+    //TODO-HACK: This is a terrible horrible ugly hack that should not be merged
+    context: std::sync::Arc<std::sync::Mutex<Vec<std::sync::Arc<&'static mut RunContext<'static, 'static, 'static>>>>>,
```
Ah yeah.  I don't like these extra args.  The hack that really upset me was this one: https://github.com/luketpeterson/hyperon-experimental/blob/b8ccce3158ff558a65ee5e61d201c15d5331e23a/lib/src/metta/runner/mod.rs#L573

But they are actually all part of the same issue.

luketpeterson commented on lib/src/metta/runner/arithmetics.rs:
```diff
@@ -107,7 +107,9 @@ impl Grounded for Bool {
+        //TODO: THis is a stop-gap until we have a more robust value-bridging system in place
+        // https://github.com/trueagi-io/hyperon-experimental/issues/351
+        match_by_string_equality(&self.to_string(), other)
     }
```
I ended up addressing this in https://github.com/trueagi-io/hyperon-experimental/pull/580/commits/5b46301c8b32afa99c8f182a61fb8716aa9c700b  The fix has the advantage that it also addresses a rust grounded atom trying to match a python atom, instead of just the reverse.

All that being said, I'm still a little uneasy about the performance of wrapping simple python objects like a bool with a grounded atom.

Alexey has a good point about not automatically down-converting when we consider big python types or types with differing behaviors (`int` being the primary example), but the cost of crossing the python bridge to check a bool doesn't sit well with me.

I would like to address this in the future.

luketpeterson commented:
To be clear, even though there is more work I plan to do before I consider this task finished, I think it should not effect the interface with the possible exception of https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1491332304

Otherwise we can make this merge to unblock others and I can continue to iterate in the background.

luketpeterson commented:
This comment https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1491205886
 (and several other warts in the design) are now cleaned up in 8a0798e
